File: 327 Count of Range Sum.java
/*
lt     mid      rt
0 1 2 3 4 5 6 7
*/
public int merge(long[] arr, long[] buf, int lt, int mid, int rt, int lower, int upper){
    int lo = mid; // first index arr[lo] - arr[i] >= lower
    int up = mid; // first index arr[up] - arr[i] > upper
    int inv = 0;

    for(int i = lt, j = mid, b = lt;i < mid;i++){
        long vi = arr[i];
        
        while(lo < rt && arr[lo] - vi <  lower) lo++;
        while(up < rt && arr[up] - vi <= upper) up++;
        inv += up - lo;
        
        while(j < rt && arr[j] <= vi) buf[b++] = arr[j++];
        buf[b++] = vi;
    }
    
    System.arraycopy(buf,lt,arr,lt,rt-lt);
    println(inv);
    return inv;
}

public int countRangeSum(int[] nums, int lower, int upper) {
    int length = nums.length;
    long[] arr = new long[length+1];
    long s = 0;
    for(int i = 0;i<length;i++){
        s += nums[i];
        arr[i+1] = s;
    }
    length++;
    
    long[] buf = new long[length];
    System.arraycopy(arr,0,buf,0,length);
    
    int inv = 0;
    int bound = length * 2;
    for(int stride = 2; stride < bound; stride <<= 1){
        for(int lt = 0;lt<length;lt += stride){
            int mid = Math.min(lt + (stride >> 1),length);
            int rt  = Math.min(lt + stride,length);
            inv += merge(arr,buf,lt,mid,rt,lower,upper);
        }
    }
    return inv;
}

File: 327 Count of Range Sum.py
'''
327. Count of Range Sum
https://leetcode.com/problems/count-of-range-sum/

Given an integer array nums, return the number of range sums that lie in [lower, upper] inclusive.
Range sum S(i, j) is defined as the sum of the elements in nums between indices i and j (i ≤ j), inclusive.

Example:

Input: nums = [-2,5,-1], lower = -2, upper = 2,
Output: 3 
Explanation: The three ranges are : [0,0], [2,2], [0,2] and their respective sums are: -2, -1, 2
'''
def mergesort(arr, lower, upper, left, right):
    if left >= right:
        return 0

    mid = left + (right - left) // 2
    count = mergesort(arr, lower, upper, left, mid) + mergesort(arr, lower, upper, mid + 1, right)

    # Count of range sum
    midBegin = mid + 1
    lo, hi = midBegin, midBegin
    for e in arr[left : midBegin]:
        # This is to count how many sum ranges smaller than lower
        while lo <= right and arr[lo] - e < lower:
            lo += 1

        # This is to count how many sum ranges smaller or equal to upper
        while hi <= right and arr[hi] - e <= upper:
            hi += 1

        count += hi - lo

    # Use a lasy simple sort to merge
    arr[left : right+1] = sorted(arr[left : right+1])
    return count

class Solution:
    def countRangeSum(self, nums: List[int], lower: int, upper: int) -> int:
        # Build accumulation array
        arr = [0] * (len(nums)+1)
        for i in range(len(nums)):
            arr[i+1] = arr[i] + nums[i]

        return mergesort(arr, lower, upper, 0, len(arr)-1)
